presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java

```diff
@@ -243,6 +249,16 @@ private boolean queryCreated(DispatchQuery dispatchQuery)
             stats.trackQueryStats(dispatchQuery);
         }
 
+        // capture failures which happen before query is succesfully dispatched.
+        dispatchQuery.addStateChangeListener(newState -> {
+            if (newState == FAILED) {
+                if (!dispatchQuery.getDispatchedFuture().isDone() || !getUnchecked(dispatchQuery.getDispatchedFuture())) {
```

Comment:
can it be `! done` just _yet_?

is here a race condition which may result in double notification?

Reply:
I did not want to depend on listener execution order. 
Specifically with:
https://github.com/prestosql/presto/blob/bca1470858a7255e7e5e469e9bc019bea311f402/presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java#L88-L93

Which actually change the state of `dispatchQuery.getDispatchedFuture`.

Technically maybe this listener I add here could be merged with one I reference ...